File: src/types/transcript.ts
export type SegmentType = "Speech" | "Noise" | "Music" | "Silence" | "Unknown";

export interface TokenConfidence {
  token: string;
  confidence: number;
}

export interface TranscriptSegment {
  id: string;
  start_ms: number;
  end_ms: number;
  text: string;
  speaker_id: string | null;
  speaker_name: string | null;
  confidence: number;
  token_confidences: TokenConfidence[];
  is_flagged: boolean;
  flag_reason: string | null;
  segment_type: SegmentType;
}

export interface Speaker {
  id: string;
  name: string | null;
  color: string;
}

export interface Transcript {
  id: string;
  source_file: string;
  audio_path: string;
  segments: TranscriptSegment[];
  speakers: Speaker[];
  created_at: string;
  duration_ms: number;
}

export interface PipelineOptions {
  whisper_model: string;
  enable_llm_review: boolean;
  language: string | null;
  confidence_threshold: number;
  hf_token: string | null;
  ollama_model: string | null;
  ollama_url: string | null;
}

export interface PipelineProgress {
  step: string;
  percent: number;
  message: string;
}

export interface ExportOptions {
  format: "Txt" | "Md" | "AiMd";
  include_timestamps: boolean;
  include_speaker_labels: boolean;
  highlight_low_confidence: boolean;
  include_annotations: boolean;
}

export type ChipTier = "Base" | "Pro" | "Max" | "Other";

export interface HardwareInfo {
  ram_gb: number;
  chip: string;
  chip_tier: ChipTier;
  gpu_cores: number | null;
  model_name: string;
}

export type ProfileCategory = "Whisper" | "Ollama";

export interface ModelProfile {
  category: ProfileCategory;
  id: string;
  name: string;
  quality_stars: number;
  speed_label: string;
  estimated_realtime_factor: string;
  ram_required_gb: number;
  shortcomings: string[];
  available: boolean;
}

File: src-tauri/src/models/transcript.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum SegmentType {
    Speech,
    Noise,
    Music,
    Silence,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenConfidence {
    pub token: String,
    pub confidence: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TranscriptSegment {
    pub id: String,
    pub start_ms: u64,
    pub end_ms: u64,
    pub text: String,
    pub speaker_id: Option<String>,
    pub speaker_name: Option<String>,
    pub confidence: f32,
    pub token_confidences: Vec<TokenConfidence>,
    pub is_flagged: bool,
    pub flag_reason: Option<String>,
    pub segment_type: SegmentType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Speaker {
    pub id: String,
    pub name: Option<String>,
    pub color: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Transcript {
    pub id: String,
    pub source_file: String,
    /// Absolute path to the 16kHz mono WAV produced during the pipeline.
    /// Used by the frontend to play back audio for each segment.
    pub audio_path: String,
    pub segments: Vec<TranscriptSegment>,
    pub speakers: Vec<Speaker>,
    pub created_at: String,
    pub duration_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiarizationSegment {
    pub start: f64,
    pub end: f64,
    pub speaker: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineOptions {
    pub whisper_model: String,
    pub enable_llm_review: bool,
    pub language: Option<String>,
    pub confidence_threshold: f32,
    pub hf_token: Option<String>,
    pub ollama_model: Option<String>,
    pub ollama_url: Option<String>,
}

impl Default for PipelineOptions {
    fn default() -> Self {
        Self {
            whisper_model: "large-v3".to_string(),
            enable_llm_review: true,
            language: None,
            confidence_threshold: 0.7,
            hf_token: None,
            ollama_model: Some("llama3.1:8b".to_string()),
            ollama_url: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineProgress {
    pub step: String,
    pub percent: u8,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExportOptions {
    pub format: ExportFormat,
    pub include_timestamps: bool,
    pub include_speaker_labels: bool,
    pub highlight_low_confidence: bool,
    pub include_annotations: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ExportFormat {
    Txt,
    Md,
}

impl Default for ExportOptions {
    fn default() -> Self {
        Self {
            format: ExportFormat::Md,
            include_timestamps: true,
            include_speaker_labels: true,
            highlight_low_confidence: true,
            include_annotations: true,
        }
    }
}
